wasi-nn.wit.md

```diff
@@ -0,0 +1,137 @@
+type tensor-dimensions = list<u32>
+
+// The type of the elements in a tensor.
+enum tensor-type {
+    fp16,
+    fp32,
+    up8,
+    ip32
+}
+
+// The tensor data.
+//
+// Initially coneived as a sparse representation, each empty cell would be filled with zeros and
+// the array length must match the product of all of the dimensions and the number of bytes in the
+// type (e.g., a 2x2 tensor with 4-byte f32 elements would have a data array of length 16).
+// Naturally, this representation requires some knowledge of how to lay out data in memory--e.g.,
+// using row-major ordering--and could perhaps be improved.
+type tensor-data = list<u8>
+
+// A tensor.
+record tensor {
+    // Describe the size of the tensor (e.g., 2x2x2x2 -> [2, 2, 2, 2]). To represent a tensor
+    // containing a single value, use `[1]` for the tensor dimensions.
+    dimensions: tensor-dimensions,
+
+    // Describe the type of element in the tensor (e.g., f32).
+    tensor-type: tensor-type,
+
+    // Contains the tensor data.
+    data: tensor-data,
+}
+```
+
+### Graph
+
+A `graph` is a `load`-ed instance of a specific ML model (e.g., MobileNet) for a specific ML
+framework (e.g., TensorFlow):
+
+```wit
+// The graph initialization data.
+//
+// This consists of an array of buffers because implementing backends may encode their graph IR in
+// parts (e.g., OpenVINO stores its IR and weights separately).
+type graph-builder = list<u8>
+type graph-builder-array = list<graph-builder>
+
+// An execution graph for performing inference (i.e., a model).
+resource graph
+
+// Describes the encoding of the graph. This allows the API to be implemented by various backends
+// that encode (i.e., serialize) their graph IR with different formats.
+enum graph-encoding {
+    openvino,
+    onnx,
+    tensorflow
+}
+
+// Define where the graph should be executed.
+enum execution-target {
+    cpu,
+    gpu,
+    tpu
+}
+
+// Load an opaque sequence of bytes to use for inference.
+load: function(builder: graph-builder-array, encoding: graph-encoding, target: execution-target) -> expected<graph, error>
+```
+
+### Inference
+
+An inference "session" is encapsulated by a `graph-execution-context`. This structure binds a
+`graph` to input tensors before `compute`-ing an inference:
+
+```wit
+//
+resource graph-execution-context
+
+// Create an execution instance of a loaded graph.
+init-execution-context: function(graph: graph) -> expected<graph-execution-context, error>
+```
+
+Inputs and outputs to the inference request &mdash; `compute` &mdash; are identified by a numeric
+index:
+
+```wit
+// Define the inputs to use for inference.
+set-input: function(ctx: graph-execution-context, index: u32, tensor: tensor) -> expected<_, error>
+
+// Compute the inference on the given inputs.
+compute: function(ctx: graph-execution-context) -> expected<_, error>
+
+// Extract the outputs after inference.
+get-output: function(ctx: graph-execution-context, index: u32) -> expected<tensor, error>
```

Comment:
how is this expected to work?
wit-bindgen seems to produce code to only return pointers to data, not the data itself, for c.
rust-wasm seems similar. (but i might be wrong because i'm not familiar with rust.)

Reply:
I think the missing piece here is [`realloc` in the canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md). `realloc` provides a way for the host to allocate for itself enough space in the guest memory in which to place the tensor bytes.